### PR TITLE
BUG: Remove null values before sorting during groupby nunique calculation

### LINKED ISSUE #27904: groupby nunique() with dates vs datetimes in presence of NaTs
#### Code Sample

```python

import numpy as np
import pandas as pd
import datetime

df = pd.DataFrame({'a': ['x','x','x','x','x'], 'b': [datetime.date(2019,1,1), pd.NaT, datetime.date(2019,1,1), pd.NaT, datetime.date(2019,1,1)]})
print(df)

# unique() returns only 1 value (plus the NaT) as expected
print('unique() with dates')
print(df.groupby(['a'])['b'].unique())

# nunique() returns 3 instead of the expected 1 (NaTs are dropped). Seems to be due to the NaTs
print('nunique() with dates')
print(df.groupby(['a'])['b'].nunique())

# convert date column to datetime
df['b'] = df.b.astype(np.datetime64)

# with datetime, unique() still returns 1 value (plus the NaT) as expected
print('unique() with datetimes')
print(df.groupby(['a'])['b'].unique())

# with datetime, nunique() returns 1 as expected
print('nunique() with datetimes')
print(df.groupby(['a'])['b'].nunique())

```
#### Problem description
`Series.groupby().nunique()` seems to have a different behavior with dates vs datetimes when NaTs are in the Series. The behavior with `date` seems to return more distinct values than there are.
This is not the same issue as https://github.com/pandas-dev/pandas/issues/14149

#### Expected Output
For nunique() with dates the expected output is:
```
a
x    1
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.18.0-25-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.0
numpy            : 1.16.4
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.2
setuptools       : 41.0.1
Cython           : 0.29.13
pytest           : 5.0.1
hypothesis       : None
sphinx           : 2.1.2
blosc            : None
feather          : None
xlsxwriter       : 1.1.8
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.7.0
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.6.9
odfpy            : None
openpyxl         : 2.6.2
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.6
tables           : 3.5.2
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.1.8

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] Closes #27904
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -120,6 +120,8 @@ Groupby/resample/rolling
 - Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.transform` where applying a timezone conversion lambda function would drop timezone information (:issue:`27496`)
 - Bug in windowing over read-only arrays (:issue:`27766`)
 - Fixed segfault in `pandas.core.groupby.DataFrameGroupBy.quantile` when an invalid quantile was passed (:issue:`27470`)
+- Bug in :meth:`pandas.core.groupby.SeriesGroupBy.nunique` where ``NaT`` values were interfering
+with the count of unique values
```
Comment (jschendel) [inline doc/source/whatsnew/v0.25.1.rst]: Can you add a reference to the issue number?

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1143,6 +1143,11 @@ def nunique(self, dropna=True):
 
         val = self.obj._internal_get_values()
 
+        if dropna:
+            mask = np.where(val == val)[0]
```
Comment (jschendel) [inline pandas/core/groupby/generic.py]: this should be `mask = notna(val)`

File: pandas/tests/groupby/test_function.py [hunk 3]
```diff
@@ -1015,6 +1025,45 @@ def test_nunique_with_timegrouper():
...
+                        dt.date(2019, 1, 1),
+                    ],
+                }
+            ),
+            1,
+        ),
+    ],
+)
+def test_nunique_with_NaT(data, expected):
+    assert data.groupby(["key"])["data"].nunique()[0] == expected
```
Comment (jschendel) [inline pandas/tests/groupby/test_function.py]: Can you add the issue number as a comment?

Also, maybe a little more readable to explicitly define a `result`, then `assert`:
```python
result = data.groupby(["key"])["data"].nunique()[0]
assert result == expected
```

Comment (MarcoGorelli) [reply pandas/tests/groupby/test_function.py]: Sure, thanks for your feedback!

**Comment by TomAugspurger** [comment]
Does this completely close #27904? Or should it remain open for the case of `dropna=False`?

**Comment by MarcoGorelli** [comment]
In the case of `dropna=False`, what should `nunique` return? As in, do the `NaN`s all count as distinct values? I wouldn't expect them to, although as `np.nan == np.nan` returns `False`, I'm not sure

**Comment by TomAugspurger** [comment]
Just one NA value I think.

**Comment by TomAugspurger** [comment]
And to be clear, we can leave handling `dropna=False` to a followup PR. Just want to make sure that we have an open issue for it.

**Review by TomAugspurger — APPROVED** [review]
This looks good for the dropna=True case. Will leave the issue open for dropna=False, which seems broken still.

**Comment by TomAugspurger** [comment]
Actually... I'm going to push this off the 0.25.1 milestone (sorry @MarcoGorelli)

Can you check a few things:

1. How's the performance of this? I assume `np.lexsort` has to do some kinds of NA masking and filtering internally. Are we much slower than that?
2. It looks like there's a now unnecessary `if dropna` starting around line 1171 now. IIUC, `val` should now have all the na values removed, so masking again shouldn't be necessary.

**Comment by MarcoGorelli** [comment]
Sorry for the delay - yes, I agree with your decision, Tom :) I'll pick this up again next week

**Comment by MarcoGorelli** [comment]
As noted as a comment in #27904, this works fine with numpy's own `nat` value...is it considered cheating just replace `pandas.NaT` with `np.datetime64("NaT")`? I tried that, and both the existing and the new tests pass.

The only test that fails is `test_quantile_out_of_bounds_q_raises`, but as far as I can tell that fails on the master branch as well, so it's not caused by my change.

**Comment by WillAyd** [comment]
Haven't looked deeply yet but does this also impact Series and DataFrame methods? The location of the change would imply so hence my reason for asking

**Review by WillAyd — CHANGES_REQUESTED** [review]
See above comment. Can you also add a whatsnew for v1.0.0

**Comment by MarcoGorelli** [comment]
Thanks for your review!

Sure, whatsnew entry added.

If I've understood your question, then the `pandas.DataFrame.nunique` and `pandas.Series.nunique` didn't have this issue to begin with, and their methods are unaffected by this change.

**Comment by WillAyd** [comment]
Ah misread the file location but makes sense so thanks for confirming. Should be able to review more deeply over the next few days

File: pandas/core/groupby/generic.py [hunk 4]
```diff
@@ -1143,6 +1143,9 @@ def nunique(self, dropna=True):
 
         val = self.obj._internal_get_values()
 
+        # GH 27951
+        val[isna(val)] = np.datetime64("NaT")
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: I think the actual root of the issue is a bug in NumPy as described by @TomAugspurger where NaT values are not sorted as you'd expected

https://github.com/numpy/numpy/issues/12629

So I think this works for now but maybe add a comment about NumPy bug 12629 for reference

File: pandas/tests/groupby/test_function.py [hunk 5]
```diff
@@ -1015,6 +1025,81 @@ def test_nunique_with_timegrouper():
     tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.parametrize(
```
Comment (WillAyd) [inline pandas/tests/groupby/test_function.py]: The parametrization here is pretty repetitive, though I realize that you have three items at a time being sent through to keep the expectation different across each.

Is there a way to more succinctly parametrize though? It's rather difficult to read this and find what's expected

**Review by WillAyd — CHANGES_REQUESTED** [review]
sorry for delay in review. Looks pretty good some general comments

File: pandas/tests/groupby/test_function.py (hunk 5 above)
Comment (MarcoGorelli) [reply pandas/tests/groupby/test_function.py]: Agreed - have modified it so the DataFrame is constructed within the test

**Review by WillAyd — APPROVED** [review]
lgtm - @TomAugspurger mind another quick look?

**Comment by TomAugspurger** [comment]
Thanks @MarcoGorelli!

Can you confirm that this fixed all of #27904, or are there outstanding tasks?

**Comment by MarcoGorelli** [comment]
That's right - thanks for having taken the time to review my work!

**Pull request merged**